### ISSUE TITLE
Multiple subkeys created for my PGP key

### ISSUE THREAD

**Issue by kirsle** [keybase/keybase-issues#2610]
I allowed Keybase to generate its own 4096-bit key on the website, and I later exported it (and its private key) to have in my local gpg client where I then published it to a key server.

I noticed I have *two* subkeys under my primary key: [E2E71CDA](https://pgp.mit.edu/pks/lookup?op=vindex&search=0x925542EDE2E71CDA). Usually people only have one subkey at a time? Or when they have two, one is used for signing and one is for encrypting.

I also added one device and generated one paper key. Does adding a new device also add a new subkey? I don't want to end up with a dozen subkeys; two already seems to be too much, but I don't even know which one is the "primary" subkey so I can revoke the other one, or if that would break Keybase if I did.

## tl;dr.:

* Why does my PGP key have two subkeys?
* Which subkey gets used when somebody encrypts a message to me?
* Can I revoke one of them?
* Does adding a new device add a new subkey (5 devices = 5 new subkeys)?

**Comment by juanminm** [comment]
If that's true, it looks like Keybase does advanced way when making a GnuPG key. But I can see it being confusing or hard for new GnuPG users. I can't use Keybase at the moment so I can't test it if it really does it.

**Glossary:**
C = Certification
S = Signing
E = Encryption
A = Authentication
 
> Why does my PGP key have two subkeys?

**Short:** Keybase seems to have created for you a master key with SC flags with two subkeys: a subkey with SA flags and a subkey with E flags.

**Long:** Because on a secure setup, you want to have your master key with only the Certification (C) usage for creating new subkeys and signing/certifying other keys and keep that master key off an online system (either on an air gapped computer or a OpenPGP smartcard). Even if you don't use the offline setup, using S, E or A subkeys instead of a master key is usually the best scenario when a device gets compromised as you only need to revoke that subkey instead of the master key.

I haven't read the source code for Keybase or the OpenPGP Go library to see if Keybase generates a master key with SC flags and the two subkeys with SA and E flags respectively, nor I'm on a system with Keybase right now to test that. GnuPG's default is to create a 2048-bit RSA key with SC flags and a 2048-bit RSA subkey with E flag.

> Which subkey gets used when somebody encrypts a message to me?

As you only have an E subkey flag (DA1B078D), it will use that subkey for encrypting at you.
The SA subkey (48999AEA) will only be used when you sign messages/files or want to authenticate via SSH.
The SC master key won't be used for signing unless the SA subkey is missing, but it will be used for creating subkeys and signing/certifying public keys.

> Can I revoke one of them?

Yes, if you don't need the SA subkey for signing messages/files or authenticating via SSH, you can revoke that subkey. It shouldn't affect anything but keep a back-up with the non-revoked subkey and don't upload it into a keyserver until you've noticed nothing faulty.

> Does adding a new device add a new subkey (5 devices = 5 new subkeys)?

It shouldn't.

> I don't want to end up with a dozen subkeys; two already seems to be too much, but I don't even know which one is the "primary" subkey so I can revoke the other one, or if that would break Keybase if I did

The only thing that could break Keybase would be revoking the master key, if it ever breaks it.

There is no primary subkey except for encryption where it gets complicated as you have to know to which subkey it was encrypted to (normally the most recently created or updated non-expired and non-revoked one unless specified by you or the user encrypting it to you). You can have multiple signing subkeys (example: one per real-life device: desktop, smartphone, work desktop, etc...) and people will verify the messages from any of them correctly.
If you have more than one signing subkey in the same device it will use only the most recently created or updated non-expired and non-revoked one, unless you specify it. It will only fallback to the master key when there is no signing subkey available.

**Comment by kirsle** [comment]
Thanks! I'm pretty new to GPG and decided Keybase was a good excuse to get into it and was just making sure it didn't generate weird keys for me. Everything you said makes sense with the official GPG docs too (having different subkeys for signing vs. encrypting).
